### PR TITLE
Add MOAB skinner

### PR DESCRIPTION
**Pull request by aprilnovak**
This copies the `MoabUserObject` from [Aurora](https://github.com/aurora-multiphysics/aurora) into Cardinal to help streamline our efforts to couple Aurora and Cardinal.

This skinner will group elements according to (i) subdomain, (ii) temperature, and (iii) density and re-generate MOAB surface meshes that bound these unique regions, like so:

Screen Shot 2023-01-27 at 3 53 40 PM

### Summary of differences from Aurora's implementation

- Choose the density bins based on an absolute spacing (as opposed to a percent difference). This eliminates rigid reliance on a particular class (`OpenMCDensity`) - there's nothing technically wrong with the user setting density with a `GenericConstantMaterial` or similar object that doesn't have an `initialDensity()` function. 
- Choose the "material" bins according to subdomain IDs, rather than material names. I think this should be functionally equivalent, without _needing_ to have material objects exist. In Cardinal, we set density in OpenMC by reading from a density `AuxVariable` instead of a density `Material`. I think this is more flexible and also allows density to be a solution variable from another app (e.g. Nek).
- Allow `MoabSkinner` to be entirely independent of another object "driving" it. This allows me to make these tests separate from running OpenMC, which is nice for modularization.
- Renamed `MoabUserObject` -> `MoabSkinner` in case we ever want to copy verbatim Aurora's objects into Cardinal, to avoid a name collision. I also think `MoabUserObject` is a nice base class name, if we were want to have different skinner daughter classes.
- Look up variable values directly from the serialized solution, which eliminates all the stuff related to the `MeshFunctions` (and should be faster I think).
- Remove the error catching from MOAB calls, because I'm not super clear on whether any of those interface calls would actually ever fail. This makes the code easier to read.

### New features

- `SkinnedBins` auxkernel allows you to see which bin each element maps to, on the original volume Exodus mesh. If there's a lot of bins, then the skins output with `moab_skins_<n>.h5m` can be hard to look at in Paraview.
- Added many regression-type tests, which check (i) validity of input parameters, (ii) tetrahedral mesh input, (iii) binning operations, and (iv) `.h5m` file write options.
- The logic for binning was moved into a `BinUtility` class, because some other user objects in Cardinal did basically the same operations already (for the Nek binning user objects).
- New documentation: [MoabSkinner](https://cardinal.cels.anl.gov/source/userobjects/MoabSkinner.html), [SkinnedBins](https://cardinal.cels.anl.gov/source/auxkernels/SkinnedBins.html)

### Remaining items

This PR does not actually skin OpenMC geometries using the re-generated DAGMC surfaces. That will come in a later PR.

### CODE REVIEW

**Comment by aprilnovak** [comment]
@helen-brooks I added you to the cardinal-devs organization, keep an eye out for an invitation

File: doc/content/source/auxkernels/SkinnedBins.md
Comment (pshriwise) [inline doc/content/source/auxkernels/SkinnedBins.md:11]: If we skin the tet mesh by temperature only, the temperature regions may span multiple subdomains (i.e. materials). In that case, how are materials assigned in the DAGMC model?

Comment (aprilnovak) [reply doc/content/source/auxkernels/SkinnedBins.md:11]: This comment here for `SkinnedBins` only refers to visualizing the bins (so you can separately look at the temperature bins, the density bins, and the subdomain bins -- or the convolution of them all).

In terms of the skinning operation (in `MoabSkinner`), we always skin by temperature + subdomain. So the mesh subdomains need to correspond to material boundaries in the DAGMC model. Eventually I'd like to add some error checking to enforce that, but I plan to leverage the mesh <--> cell mapping in `OpenMCCellAverageProblem` (which already has functions to map cell instances/materials to mesh elements) to do that. So I was going to save it for a future PR.

File: src/userobjects/MoabSkinner.C
Comment (pshriwise) [inline src/userobjects/MoabSkinner.C:849]: I think this can be a `const` reference to the vector to avoid an unecessary memory allocation.

File: src/userobjects/MoabSkinner.C
Comment (pshriwise) [inline src/userobjects/MoabSkinner.C:923]: Pretty sure this one can be `const` as well.

File: src/userobjects/MoabSkinner.C
Comment (pshriwise) [inline src/userobjects/MoabSkinner.C:986]:
```suggestion
  // Create each tri by a cyclic permutation of indices
  // Values of i1, i2 in the loop: 0,1; 1,2; 2;0
  for (unsigned int i = 0; i < 3; i++)
  {
```

File: src/userobjects/MoabSkinner.C
Comment (pshriwise) [inline src/userobjects/MoabSkinner.C:1000]:
```suggestion
MoabSkinner::createTri(const std::vector<moab::EntityHandle> & vertices,
                       unsigned int v1,
                       unsigned int v2,
                       unsigned int v3)
```

**Review by pshriwise — CHANGES_REQUESTED** [review]
This looks really nice!! Very excited to see what we can do by combining these capabilities. Some minor comments from me here as an initial pass.

File: include/userobjects/MoabSkinner.h
Comment (aprilnovak) [reply include/userobjects/MoabSkinner.h:219]: I don't think it's important, they will get rendered the same by Doxygen. The `/** **/` one is just necessary for multi-line comments, I think.

File: src/userobjects/MoabSkinner.C
Comment (aprilnovak) [reply src/userobjects/MoabSkinner.C:849]: This gets pushed_back to later

File: src/userobjects/MoabSkinner.C
Comment (aprilnovak) [reply src/userobjects/MoabSkinner.C]: Yep!

File: src/userobjects/MoabSkinner.C
Comment (pshriwise) [reply src/userobjects/MoabSkinner.C:845]: Whoops, my bad!

File: doc/content/source/auxkernels/SkinnedBins.md
Comment (helen-brooks) [reply doc/content/source/auxkernels/SkinnedBins.md:11]: Good question, and helpful answer.
As for equivalency with DAGMC model I would advocate for eliminating the dagmc model step entirely. I have an open branch in AURORA (unfortunately not touched since the summer) which will remove the need for it.

Comment (aprilnovak) [reply doc/content/source/auxkernels/SkinnedBins.md:11]: Nice! So you mean that the initial DAGMC model would be built straight from the MOOSE input files, with no need for a .h5m file?

Comment (aprilnovak) [reply doc/content/source/auxkernels/SkinnedBins.md:11]: That will be a really nice feature. For now, this PR doesn't make any assumptions about OpenMC, since the skinning "operation" is separated. So we can certainly incorporate those developments in a later PR (without conflicting with anything here, is what I'm meaning to emphasize).

File: src/userobjects/MoabSkinner.C
Comment (aprilnovak) [reply src/userobjects/MoabSkinner.C:580]: Yeah, I'd like to have an option to truncate in the future. I think I have a similar to-do item in OpenMC (in case you went like 1K over the loaded XS data sets). I think for now, I'd like to keep it as an error, and add an option like `truncate = true` to throw a warning instead (in a future PR). What do you think?

File: include/userobjects/MoabSkinner.h
Comment (aprilnovak) [reply include/userobjects/MoabSkinner.h:372]: For most variables, I just kept whatever name Helen originally used (but with a pre-pended underscore to follow the MOOSE convention for denoting member variables) - but definitely open to new names! I can make this change.

File: src/userobjects/MoabSkinner.C
Comment (aprilnovak) [reply src/userobjects/MoabSkinner.C:271]: You mean wrap this in an openmp loop?

File: src/userobjects/MoabSkinner.C
Comment (aprilnovak) [reply src/userobjects/MoabSkinner.C:413]: I'm not sure exactly what you're suggesting - are you saying that "3" should be swapped with `GEOM_DIMENSION`?

File: src/userobjects/MoabSkinner.C
Comment (aprilnovak) [reply src/userobjects/MoabSkinner.C:816]: It's here for future capability. This is basically the function that is called to encode a region's material name, after the skinning operation. The function looks simpler than it is in Aurora, because in Aurora a new material is created even when only skinning by temperature (which isn't strictly needed, because you can change temperatures of cells filled with the same material without needing to duplicate the material).

**Comment by helen-brooks** [comment]
Hello so this is not yet a full review - I've not had time to go through the tests or the "SkinnedBins", but I've already got a lot of feedback just for the MOAB Skinner.

## Removed Features
 I've noticed that you've removed a lot of functionality relative to AURORA with your changes and I wonder how intentional that was. I'll list them, and perhaps you could comment on why this functionality  removed?

### Major features removed
- Support for nodal temperature and density  variables (or anything not const monomial)
  (coming from the removal of mesh function, and simply using the dof of an elem to look  the temperature result
- Support for log binning
- Support for user-setable lengthscales between MOAB mesh and MOOSE mesh. I appreciate you'd want the default to be unity, but I still thing it should be possible to scale between e.g. m and cm units
- Support for displaced mesh via the change in how the mesh is looked up.
- Support for second order / refined mesh?? (Not sure)

### Minor features removed 
- Restriction to using auxiliary variables (by use of the method getAuxiliaryVariableNumber, rather than first looking up the libMesh system number)
- It was previously possible to produce output in a periodic way, say every 5 timesteps  - which has the value if you want to produce some output but not slow down your run by writing every time.
- Performance timers

## Other changes I'm not sure about / slightly concerned by

### Major 
- I'm concerned that change in how the mesh elements and nodes are iterated over is different  - it looks like it's only serial now? And I'm not sure if it'll work for second order mesh elements or refined elements? 

### Minor
- Bin look up - the bin utility queries the value against the bounds of each bin, which to me is inefficient when you can compute the bin directly, and the integer returned will immediately also tell you if it is out of bounds (i.e. any value not between [0,nBins-1]
- Reconciliation of initialize and initMOAB functions. If memory serves me correctly, I overrided initialize to do nothing for a reason, it happened at the wrong point in time - there were unitialized pointers I wanted to use which were not available. Anyway - might need to look at this in the context of a multiphysics problem, i.e. think ahead to how the object will be used
- Removal of use of MOOSE material block names. I can envisage a scenario where a mesh might contain inactive blocks for the purpose of the neutronic/thermal solve - e.g. if we have vacuum mesh, which might happen for EM simulations. There's a real use case of impact of neutron heating on superconducting field coils...
- Removal of checking MOAB rvals. This concerns me. I think it's important to have trust barriers... It's not for me to understand every inner working of MOAB, but its rval code contains failure information. If I accidentally hand garbage to MOAB and it fails, I want to know as early as possible. If you really don't want to propagate a bunch of booleans, then I think we should add in MooseErrors or some other exception for non-zero rval.
- There seems to be a lot less error checking generally. I'm a fan of coding my assumptions, so that if there is a bug or edge case I've not thought of it is at least obvious where it's going wrong, and not propagating garbage / nonsense results and failing silently.

## Some changes I thought were very nice
- Use of  checkRequiredParam, addRangeCheckedParam and other user input sanitization
- Style improvements, like more consistent use of snake case and "_" for internal variables
- Ability to execute as "standalone" - presumably that's to make the MOOSE regression tests work 

That's probably enough to discuss for now...

**Comment by aprilnovak** [comment]
Thanks @helen-brooks for the really detailed review! I'll try to address your points in the same order. Going to split into separate comments for cleanliness.

- > Support for nodal temperature and density variables

This was intentional. Actually, I believe that the implementation in this PR is functionally equivalent to what is currently in Aurora. In Aurora, the `meshFunctions` are passed the element centroid. Conversely, in Cardinal we already impose that the temperature and density are constant monomials in the `openmc.i`. These variables can be _any_ other basis as-needed for the thermal-fluid solve, they just get mapped to a constant monomial basis with a multiapp transfer. I believe the `MultiAppNearestNodeTransfer`/etc. actually do that mapping with mesh functions, that assign the constant monomial DOFs (which are at the element centroid) by querying the higher-order "source" app's data. So the only difference here is whether (i) Aurora is responsible for doing that mapping, vs. (ii) we leverage MOOSE's existing transfer capabilities to keep that responsibility in the MOOSE framework.

**Comment by aprilnovak** [comment]
> Support for log binning

I removed this for now just to simplify the code. It can definitely be re-added later. I was wondering if you saw some particular cases where this was motivated because linear binning was a bad idea?

> Support for user-setable lengthscales between MOAB mesh and MOOSE mesh. I appreciate you'd want the default to be unity, but I still thing it should be possible to scale between e.g. m and cm units

Yes, this will be handled by our OpenMC wrapping. There is a `_scaling` member variable on `MoabSkinner` that is currently just set to 1.0 in the member initialization list. I am going to set it externally with the `MoabSkinner::setScaling` function.

> Support for displaced mesh via the change in how the mesh is looked up.

I'd love to have this in the future. There's other parts of Cardinal that don't support this yet though, so I wanted to leave that to a future PR.

> Support for second order / refined mesh?? (Not sure)

This is also going to be a future capability in Cardinal, and there's other systems that need to be updated for that functionality to work.

**Comment by aprilnovak** [comment]
> Restriction to using auxiliary variables (by use of the method getAuxiliaryVariableNumber, rather than first looking up the libMesh system number)

I actually think it's not possible to have nonlinear variables with an `ExternalProblem` class, because the nonlinear system gets set to a [zero-variable system](https://github.com/idaholab/moose/blob/next/framework/src/problems/ExternalProblem.C#:~:text=%2F**,NonlinearSystem%3E(*this%2C%20%22nl0%22)%3B).

> It was previously possible to produce output in a periodic way, say every 5 timesteps - which has the value if you want to produce some output but not slow down your run by writing every time.

Yes, I removed this for simplicity. Under what situations do you want to be creating the skins all the time? With the `SkinnedBins` auxkernel I created, the skins can be visualized natively in the MOOSE Exodus output.

> Performance timers

This is just a personal preference - I don't have many timers in Cardinal yet because I haven't done a lot of performance studies. Once I get there, there'll be a concerted effort to add timers where needed.

**Comment by aprilnovak** [comment]
> I'm concerned that change in how the mesh elements and nodes are iterated over is different - it looks like it's only serial now? And I'm not sure if it'll work for second order mesh elements or refined elements?

Can you point to a specific line number(s) for this? These tests run fine in parallel.

**Comment by aprilnovak** [comment]
> Bin look up - the bin utility queries the value against the bounds of each bin, which to me is inefficient when you can compute the bin directly

This code is from elsewhere in MOOSE, but we can change the functionality later if we find out it is slow. For now, this PR is really about test-driven development. Later we can change the inner behavior for performance/readability/etc. and use the test suite as a guide to produce identical output behavior.

> Reconciliation of initialize and initMOAB functions

I found some of the Aurora design confusing, because many functions would be called on every time step, but would immediately "kick out" if some boolean flag was true. I have a local branch where the skinning operation does work on an OpenMC model (through Cardinal), so I'm confident this restructuring is okay.

> Removal of use of MOOSE material block names. I can envisage a scenario where a mesh might contain inactive blocks for the purpose of the neutronic/thermal solve - e.g. if we have vacuum mesh, which might happen for EM simulations. There's a real use case of impact of neutron heating on superconducting field coils...

In Cardinal, we do allow mesh blocks to "not participate" in coupling. We control this from our OpenMC wrapper. In a future PR, I'd like to add that support as well in the `MoabSkinner`.

> Removal of checking MOAB rvals

I'm of the strong opinion that, if I can design a robust and comprehensive test suite that covers the expected use case of Cardinal, that I don't need to do a lot of internal error checking of error code return values. In MOOSE in general, the style convention is to not have pretty much any return value checking (and instead throw `mooseErrors`), so I've adopted that.

**Comment by aprilnovak** [comment]
I mostly want to emphasize that this PR is not the end-all state of the `MoabSkinner`. It is a minimum functionality that allows meaningful regression tests and reduces the burden of a much bigger PR review by compartmentalizing things. I really appreciate all the comments, and will definitely address some missing features in later PRs :)

**Comment by pshriwise** [comment]
A couple questions from me:

 >  - Support for second order / refined mesh?? (Not sure)

I didn't know that Aurora can currently convert the skin of second-order elements into a DAGMC geometry. Does MOOSE perform the tessellation of the curved faces?

> > Removal of checking MOAB rvals

>I'm of the strong opinion that, if I can design a robust and comprehensive test suite that covers the expected use case of Cardinal, that I don't need to do a lot of internal error checking of error code return values. In MOOSE in general, the style convention is to not have pretty much any return value checking (and instead throw mooseErrors), so I've adopted that.

This C-style MOAB interface has bothered me for some time. IMO there are certain operations that should be safe to perform (i.e. connectivity and coordinate queries) that shouldn't require an errorcode check. That being said, having seen the many things that can go wrong during problem setup of DAGMC problems in the past and the extra information that can be gathered from the MOAB errorcodes, they can be useful. I could see an ideal scenario in which MOAB calls during problem setup that involve metadata, topology, mesh group relationships, etc. would check error codes, but the error codes coming from more basic calls (coordinate queries and such) that are called more frequently may be ignored to save cycles. Just my 2-3 cents.

**Comment by aprilnovak** [comment]
> I could see an ideal scenario in which MOAB calls during problem setup would check error codes, but the error codes coming from more basic calls that are called more frequently may be ignored

I'm ok with a compromise - I don't know enough about these functions to know which of them have a "likelihood" to fail. It seems like there's an across-the-board design in MOAB to always return an error code, even for simple operations.

Whenever I add a `mooseError` to Cardinal, I always add an accompanying test which I know can trigger the error (otherwise, I'm a bit fuzzy on what exactly I'm protecting against).  If we can design tests where it's possible for the user to trigger bad behavior in MOAB, I'd say we definitely add the corresponding rval checks back in. Otherwise, it feels like we're testing MOAB behavior (when I think of that responsibility lying with MOAB, not Cardinal). In libMesh for instance, there are `libmesh_assert()` which only get compiled in debug mode to handle low-level error checking akin to what we're discussing. If these are hypothetical errors where we're not sure how exactly to trigger, that feels like something belonging to debug builds only (so I'd suggest `libmesh_assert` instead of `mooseError`).

**Comment by aprilnovak** [comment]
@helen-brooks @pshriwise This is ready for a new round of reviews. I added a test to cover second-order tets. For the MOAB errors, I used

```
moab::ErrorCode
MoabSkinner::check(const moab::ErrorCode input) const
{
#ifdef DEBUG
  MB_CHK_ERR(input);
#endif
  return moab::MB_SUCCESS;
}
```

File: src/userobjects/MoabSkinner.C
Comment (helen-brooks) [reply src/userobjects/MoabSkinner.C]: I think I need you to review my code Patrick.

**Review by helen-brooks — APPROVED** [review]
Happy with this, given the latest changes and our discussion off-line :-)

File: src/userobjects/MoabSkinner.C
Comment (pshriwise) [reply src/userobjects/MoabSkinner.C:287]: Meaning it might be more efficient to collect all of the data and use `_moab->create_vertices` instead of `_moab->create_vertex`. But there may be tradeoffs with that. Not something I'd worry too much about right now.

File: src/userobjects/MoabSkinner.C
Comment (pshriwise) [reply src/userobjects/MoabSkinner.C:429]: There's an additional tag that's typically set on `volume_set` that identifies the dimensionality what the elements in the mesh set are intended to represent. 

https://bitbucket.org/fathomteam/moab/src/645c5735a8fd264d68b9f8532b2b039fd350e127/src/MBTagConventions.hpp#lines-70

Comment (pshriwise) [reply src/userobjects/MoabSkinner.C:429]: It might be easier to elaborate on this offline. I can also have a look and make what I think is a mini-PR for this afterward to add this in if you prefer.

File: src/userobjects/MoabSkinner.C
Comment (pshriwise) [reply src/userobjects/MoabSkinner.C:740]: Yep! I'll make an issue to track this.

File: src/userobjects/MoabSkinner.C
Comment (pshriwise) [reply src/userobjects/MoabSkinner.C:596]: Something like this may have already been added in OpenMC for temperature interpolation: https://github.com/openmc-dev/openmc/pull/2265

I'm game to bump this to a future PR for now. Maybe create an issue and link it here so we don't lose track?

**Review by pshriwise — COMMENTED** [review]
Making a review b/c multiple comments of mine were marked as pending comments for a review and I don't know how to easily convert them 🤷🏻

File: src/userobjects/MoabSkinner.C
Comment (aprilnovak) [reply src/userobjects/MoabSkinner.C:596]: Oh nice, we can definitely take advantage of that in Cardinal. I'll make an issue to keep track of it.

Comment (aprilnovak) [reply src/userobjects/MoabSkinner.C:596]: https://github.com/neams-th-coe/cardinal/issues/234

File: src/userobjects/MoabSkinner.C
Comment (aprilnovak) [reply src/userobjects/MoabSkinner.C:429]: Sure, I'm good with tackling this in a later PR. I'll add it to this catch-all issue (https://github.com/neams-th-coe/cardinal/issues/567) to make sure we keep track of it.

**Comment by aprilnovak** [comment]
Thanks @helen-brooks and @pshriwise for the review! I'll merge this later today unless there's any objections

**Pull request merged**
